src/main/java/com/aws/greengrass/shadowmanager/sync/strategy/RealTimeSyncStrategy.java

```diff
@@ -0,0 +1,248 @@
+     */
+    public RealTimeSyncStrategy(ExecutorService executorService, Retryer retryer,
+                                RetryUtils.RetryConfig retryConfig) {
+        super(retryer, retryConfig);
+        this.syncExecutorService = executorService;
+    }
+
+    /**
+     * Starts syncing the shadows based on the strategy.
+     *
+     * @param context         an context object for syncing
+     * @param syncParallelism number of threads to use for syncing
+     */
+    @Override
+    public void startSync(SyncContext context, int syncParallelism) {
+        synchronized (lifecycleLock) {
+            if (syncing.compareAndSet(false, true)) {
+                this.context = context;
+                logger.atInfo(SYNC_EVENT_TYPE).log("Start real time syncing");
+                for (int i = 0; i < syncParallelism; i++) {
+                    syncThreads.add(syncExecutorService.submit(this::syncLoop));
+                }
+            } else {
+                logger.atDebug(SYNC_EVENT_TYPE)
+                        .log("Real time Syncing is already in progress. Ignoring request to start");
+            }
+        }
+    }
+
+    /**
+     * Stops the syncing of shadows.
+     */
+    @Override
+    public void stopSync() {
+        synchronized (lifecycleLock) {
+            if (syncing.compareAndSet(true, false)) {
+                logger.atInfo(SYNC_EVENT_TYPE).log("Stop real time syncing");
+                syncing.set(false);
+
+                logger.atDebug(SYNC_EVENT_TYPE).log("Cancel {} real time sync thread(s)", syncThreads.size());
+                syncThreads.forEach(t -> t.cancel(true));
+                syncThreads.clear();
+
+                int remaining = syncQueue.size();
+                syncQueue.clear();
+
+                if (remaining > 0) {
+                    logger.atInfo(SYNC_EVENT_TYPE)
+                            .log("Stopped real time syncing with {} pending sync items", remaining);
+                }
+            } else {
+                logger.atDebug(SYNC_EVENT_TYPE)
+                        .log("Real time Syncing is already stopped. Ignoring request to stop");
+            }
+        }
+    }
+
+
+    /**
+     * Put a sync request into the queue if syncing is started.
+     * <p/>
+     * This will block if the queue is full. This is intentional as non-blocking requires either an unbounded queue
+     * of requests, or an Executor service which creates threads from an unbounded queue.
+     * <p/>
+     * We cannot support an unbounded queue as that will lead to memory pressure - instead requests need to be
+     * throttled.
+     * <p/>
+     * Synchronized so that there is at most only one put in progress waiting to be added if queue is full
+     *
+     * @param request request the request to add.
+     * @throws InterruptedException if the thread is interrupted while enqueuing data
+     */
+    @Override
+    public void putSyncRequest(SyncRequest request) throws InterruptedException {
```

Comment:
does this method actually throw InterruptedException anymore? Since you caught it and set the interrupt on the thread I don't think it does.

Does anything calling this check if `Thread.currentThread().isInterrupted()`?

Reply:
Yeah. I agree. I refactored the code. Nothing is really throwing `InterruptedException` here.